WalletWasabi/WabiSabi/Backend/Banning/CoinVerifierApiClient.cs

```diff
@@ -1,50 +1,58 @@
 using NBitcoin;
 using Newtonsoft.Json;
-using System.Collections.Generic;
-using System.Linq;
 using System.Net;
 using System.Net.Http;
-using System.Runtime.CompilerServices;
 using System.Threading;
 using System.Threading.Tasks;
 using WalletWasabi.Logging;
 using WalletWasabi.WabiSabi.Backend.Statistics;
 
 namespace WalletWasabi.WabiSabi.Backend.Banning;
 
-public class CoinVerifierApiClient
+public class CoinVerifierApiClient : IAsyncDisposable
 {
-	public CoinVerifierApiClient(string token, HttpClient httpClient)
-	{
-		ApiToken = token;
-		HttpClient = httpClient;
-	}
+	/// <summary>Maximum number of actual HTTP requests that might be served concurrently by the CoinVerifier webserver.</summary>
+	public const int MaxParallelRequestCount = 30;
 
-	private string ApiToken { get; }
+	/// <summary>Maximum re-tries for a single API request.</summary>
+	private const int MaxRetries = 3;
 
-	private HttpClient HttpClient { get; }
-
-	public virtual async Task<ApiResponseItem> SendRequestAsync(Script script, CancellationToken cancellationToken)
+	public CoinVerifierApiClient(string apiToken, HttpClient httpClient)
 	{
+		ApiToken = apiToken;
+		HttpClient = httpClient;
+
 		if (HttpClient.BaseAddress is null)
 		{
 			throw new HttpRequestException($"{nameof(HttpClient.BaseAddress)} was null.");
 		}
+
 		if (HttpClient.BaseAddress.Scheme != "https")
 		{
 			throw new HttpRequestException($"The connection to the API is not safe. Expected https but was {HttpClient.BaseAddress.Scheme}.");
 		}
+	}
 
-		var address = script.GetDestinationAddress(Network.Main); // API provider doesn't accept testnet/regtest addresses.
+	/// <summary>Timeout for a single <see cref="SendRequestAsync(Script, CancellationToken)"/> invocation (including retries).</summary>
+	private static TimeSpan TotalApiRequestTimeout { get; } = TimeSpan.FromMinutes(3);
+
+	private string ApiToken { get; }
+
+	private HttpClient HttpClient { get; }
+
+	private SemaphoreSlim ThrottlingSemaphore { get; } = new(initialCount: MaxParallelRequestCount);
 
-		int tries = 3;
-		var delay = TimeSpan.FromSeconds(2);
+	public virtual async Task<ApiResponseItem> SendRequestAsync(Script script, CancellationToken cancellationToken)
+	{
+		var address = script.GetDestinationAddress(Network.Main); // API provider doesn't accept testnet/regtest addresses.
 
 		HttpResponseMessage? response = null;
 
-		do
+		for (int i = 0; i < MaxRetries; i++)
 		{
-			tries--;
+			// Makes sure that there are no more than MaxParallelRequestCount requests in-flight at a time.
+			// Re-tries are not an exception to the max throttling limit.
+			await ThrottlingSemaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
We should wait 300 seconds for the API request itself and no retry in case of timeout. Currently waiting here also effects the remaining time.

Reply:
> We should wait 300 seconds for the API request itself and no retry in case of timeout. Currently waiting here also effects the remaining time. 

I had a PR with the changes, but it got drifted away. Out of this PR's scope tbh. I will make a new PR for that.

Reply:
https://github.com/zkSNACKs/WalletWasabi/pull/10153/commits/8496a2533faf2b511e07c359dc24080530d05a49

Reply:
3rd party said we should wait 5 minutes for the API call.